core/src/test/java/io/confluent/kafka/schemaregistry/rest/RestApiModeTest.java

```diff
@@ -0,0 +1,131 @@
+ */
+package io.confluent.kafka.schemaregistry.rest;
+
+import org.junit.Test;
+
+import io.confluent.kafka.schemaregistry.ClusterTestHarness;
+import io.confluent.kafka.schemaregistry.avro.AvroCompatibilityLevel;
+import io.confluent.kafka.schemaregistry.avro.AvroUtils;
+import io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException;
+import io.confluent.kafka.schemaregistry.rest.exceptions.RestIncompatibleAvroSchemaException;
+import io.confluent.kafka.schemaregistry.rest.exceptions.RestInvalidSchemaException;
+import io.confluent.rest.exceptions.RestConstraintViolationException;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+public class RestApiModeTest extends ClusterTestHarness {
+
+  private static String SCHEMA_STRING = AvroUtils.parseSchema(
+      "{\"type\":\"record\","
+          + "\"name\":\"myrecord\","
+          + "\"fields\":"
+          + "[{\"type\":\"string\",\"name\":\"f1\"}]}")
+      .canonicalString;
+
+  public RestApiModeTest() {
+    super(1, true, AvroCompatibilityLevel.BACKWARD.name);
+  }
+
+  @Test
+  public void testReadOnlyMode() throws Exception {
+    String subject = "testSubject";
+    String mode = "READONLY";
+
+    // set mode to read only
+    assertEquals(
+        mode,
+        restApp.restClient.setMode(mode).getMode());
+
+    // register a valid avro schema
+    try {
+      restApp.restClient.registerSchema(SCHEMA_STRING, subject);
+      fail("Registering during read-only mode should fail");
+    } catch (RestClientException e) {
+      // this is expected.
+      assertEquals("Should get a constraint violation",
+          RestConstraintViolationException.DEFAULT_ERROR_CODE,
+          e.getStatus());
+    }
+  }
+
+  @Test
+  public void testReadWriteMode() throws Exception {
+    String subject = "testSubject";
+    String mode = "READWRITE";
+
+    // set mode to read write
+    assertEquals(
+        mode,
+        restApp.restClient.setMode(mode).getMode());
+
+    // register a valid avro schema
+    try {
+      restApp.restClient.registerSchema(SCHEMA_STRING, subject, 1, 1);
+      fail("Registering an incompatible schema should fail");
+    } catch (RestClientException e) {
+      // this is expected.
+      assertEquals("Should get a constraint violation",
+          RestConstraintViolationException.DEFAULT_ERROR_CODE,
+          e.getStatus());
+    }
+
+    int expectedIdSchema1 = 1;
+    assertEquals("Registering without id should succeed",
+        expectedIdSchema1,
+        restApp.restClient.registerSchema(SCHEMA_STRING, subject));
+  }
+
+  @Test
+  public void testImportMode() throws Exception {
+    String subject = "testSubject";
+    String mode = "IMPORT";
+
+    // set mode to import
+    assertEquals(
+        mode,
+        restApp.restClient.setMode(mode).getMode());
+
+    // register a valid avro schema
+    int expectedIdSchema1 = 1;
+    assertEquals("Registering without id should succeed",
```

Comment:
nit: I got carried away with the message. I think your testing with id but the message is incorrect :)